common/src/main/java/com/microsoft/identity/common/crypto/AndroidBrokerStorageEncryptionManager.java

```diff
@@ -0,0 +1,173 @@
+
+    protected final String LEGACY_AUTHENTICATOR_APP_KEY_ALIAS = "LEGACY_AUTHENTICATOR_APP_KEY";
+    protected final String LEGACY_COMPANY_PORTAL_KEY_ALIAS = "LEGACY_COMPANY_PORTAL_KEY";
+
+    private final Context mContext;
+    private final ITelemetryCallback mTelemetryCallback;
+    private final PredefinedKeyLoader mLegacyAuthAppKeyLoader;
+    private final PredefinedKeyLoader mLegacyCPKeyLoader;
+    private final AndroidWrappedKeyLoader mKeyStoreKeyLoader;
+
+    public AndroidBrokerStorageEncryptionManager(@NonNull final Context context,
+                                                 @Nullable final ITelemetryCallback telemetryCallback) {
+        mContext = context;
+        mTelemetryCallback = telemetryCallback;
+
+        mLegacyAuthAppKeyLoader = new PredefinedKeyLoader(LEGACY_AUTHENTICATOR_APP_KEY_ALIAS,
+                AuthenticationSettings.INSTANCE.getBrokerSecretKeys().get(AZURE_AUTHENTICATOR_APP_PACKAGE_NAME));
+
+        mLegacyCPKeyLoader = new PredefinedKeyLoader(LEGACY_COMPANY_PORTAL_KEY_ALIAS,
+                AuthenticationSettings.INSTANCE.getBrokerSecretKeys().get(COMPANY_PORTAL_APP_PACKAGE_NAME));
+
+        mKeyStoreKeyLoader = new AndroidWrappedKeyLoader(KEY_STORE_ALIAS, context, telemetryCallback);
+    }
+
+    // Exposed for Robolectric.
+    protected String getPackageName(){
+        return mContext.getPackageName();
+    }
+
+    @Override
+    public @NonNull AbstractSecretKeyLoader getKeyLoaderForEncryption() {
+        if (StorageHelper.sShouldEncryptWithKeyStoreKey) {
+            return mKeyStoreKeyLoader;
+        }
+
+        if (AZURE_AUTHENTICATOR_APP_PACKAGE_NAME.equalsIgnoreCase(getPackageName())) {
+            return mLegacyAuthAppKeyLoader;
+        }
+
+        if (COMPANY_PORTAL_APP_PACKAGE_NAME.equalsIgnoreCase(getPackageName()) ||
+                BROKER_HOST_APP_PACKAGE_NAME.equalsIgnoreCase(getPackageName())) {
+            return mLegacyCPKeyLoader;
+        }
+
+        throw new IllegalStateException("Matching encryption key not found");
+    }
+
+    @Override
+    public @NonNull List<AbstractSecretKeyLoader> getKeyLoaderForDecryption(@NonNull byte[] cipherText) {
+        final String methodName = ":getKeyLoaderForDecryption";
+        final String packageName = getPackageName();
+
+        if (isEncryptedByThisKeyIdentifier(cipherText, PredefinedKeyLoader.KEY_IDENTIFIER)) {
+            if (COMPANY_PORTAL_APP_PACKAGE_NAME.equalsIgnoreCase(packageName) ||
+                    BROKER_HOST_APP_PACKAGE_NAME.equalsIgnoreCase(packageName)) {
+                return new ArrayList<AbstractSecretKeyLoader>() {{
+                    add(mLegacyCPKeyLoader);
+                    add(mLegacyAuthAppKeyLoader);
+                }};
+            } else if (AZURE_AUTHENTICATOR_APP_PACKAGE_NAME.equalsIgnoreCase(packageName)) {
+                return new ArrayList<AbstractSecretKeyLoader>() {{
+                    add(mLegacyAuthAppKeyLoader);
+                    add(mLegacyCPKeyLoader);
```

Comment:
Why do we need CP key loader if we know the package is auth app?

Reply:
Synced offline.